### PR TITLE
issue/26384 Fix store switcher when using different base url on stores

### LINKED ISSUE #23755: Store view switcher is wrong , when each store views have different url.
### Preconditions (*)

1.  Magento 2.3.1
2.  using 2 store views and those views have different url (en.magento.local & jp.magento.local) 

### Steps to reproduce (*)

1.  Click a target store view on  language switcher.
2.  successfully switch to target store view.
3.  but will be displayed below error.
The store that was requested wasn't found. Verify the store and try again.

### Expected result (*)

1.  Switch to target store view.
2.  no errors are displayed.

### Actual result (*)

1.  Switch to target store view.
2.  below error is displayed.
The store that was requested wasn't found. Verify the store and try again.

### LINKED ISSUE #26384: Store switcher redirects to homepage for multistore setup with different domains
### Preconditions (*)

1. 2.3.3
2. More than one store view
3. Separate domain for the store views (default -> mage.local, second store view -> test.mage.local)
Apache vhost:
```
<VirtualHost *:80>
   ServerName          mage.local
   DocumentRoot        /var/www/html/magento2/pub/
</VirtualHost>

<VirtualHost *:80>
   ServerName          test.mage.local
   DocumentRoot        /var/www/html/magento2/pub/
   SetEnv MAGE_RUN_CODE "test"
   SetEnv MAGE_RUN_TYPE "store"
</VirtualHost>
```
4. Setup base urls for both store views in Stores > Settings > Configuration > General > Web
5. Create a category (it can be reproduced with cms pages too)

### Steps to reproduce (*)

1. Go to mage.local
2. Go to category
3. Use store switch to switch to the other store

### Expected result (*)

1. Redirect to test.mage.local/category-name.html

### Actual result (*)

1. Redirect to test.mage.local/

### Additional Information
Probably related to #14674 and #19714

### PR DESCRIPTION
**Pull request by TobiasLufinity**
### Description (*)
When a user switches between stores using the store switcher the request fails if the target store has a different base url than the current. This pull request changes the redirect controller to make the switch request using the targeted store' url instead so the switch is successful. 

### Fixed Issues (if relevant)
1. Fixes magento/magento2#26384 : Store switcher redirects to homepage for multistore setup with different domains
2. Fixes magento/magento2#23755 : Store view switcher is wrong , when each store views have different url.

### Manual testing scenarios (*)
1. Have simple Magento 2.4-dev setup with sampledata
2. Create two or more stores with different base url (i.e default.magento.dev and en.magento.dev)
3. Go to a category page of one store
4. Switch between the stores using the store switcher

Result without fix: You end up on the start page, without switched store and not the category on which you were on.
Result with fix: Store is switched and you stay in the same page you were on

### Questions or comments
No automated tests has been created for this pull request

### Contribution checklist (*)
 - [ ] Pull request has a meaningful description of its purpose
 - [ ] All commits are accompanied by meaningful commit messages
 - [ ] All new or changed code is covered with unit/integration tests (if applicable)
 - [ ] All automated tests passed successfully (all builds are green)

### CODE REVIEW

File: app/code/Magento/Store/Controller/Store/Redirect.php [hunk 1]
```diff
@@ -53,12 +60,14 @@ public function __construct(
         StoreResolverInterface $storeResolver,
         \Magento\Framework\Session\Generic $session,
         \Magento\Framework\Session\SidResolverInterface $sidResolver,
-        HashGenerator $hashGenerator
+        HashGenerator $hashGenerator,
+        StoreManagerInterface $storeManager
```
Comment (edenduong) [inline app/code/Magento/Store/Controller/Store/Redirect.php]: Please change it to `StoreManagerInterface $storeManager = null` for backward compatible
https://devdocs.magento.com/guides/v2.3/contributor-guide/backward-compatible-development/

File: app/code/Magento/Store/Controller/Store/Redirect.php [hunk 2]
```diff
@@ -53,12 +60,14 @@ public function __construct(
...
     ) {
         parent::__construct($context);
         $this->storeRepository = $storeRepository;
         $this->storeResolver = $storeResolver;
         $this->hashGenerator = $hashGenerator;
+        $this->storeManager = $storeManager;
```
Comment (edenduong) [inline app/code/Magento/Store/Controller/Store/Redirect.php]: Please change to`$this->storeManager = $storeManager ?? ObjectManager::getInstance()->get(StoreManagerInterface::class);` for backward compatible
https://devdocs.magento.com/guides/v2.3/contributor-guide/backward-compatible-development/

**Review by edenduong — CHANGES_REQUESTED** [review]
Thank you for your contribution. Please see my comment. And could you please cover your change by Automation Test? Thank you !

**Comment by TobiasLufinity** [comment]
Simple test made which goes through the controller and make sure that the StoreManager change to the correct store.
Also added the backwards compatibility for the controller.

File: app/code/Magento/Store/Test/Unit/Controller/Store/RedirectTest.php [hunk 3]
```diff
@@ -0,0 +1,140 @@
...
+     */
+    private $storeResolverMock;
+
+
+    /**
+     * @return void
+     */
+    protected function setUp()
+    {
+        $this->storeManagerMock = $this->getMockBuilder(\Magento\Store\Model\StoreManagerInterface::class)->getMock();
```
Comment (edenduong) [inline app/code/Magento/Store/Test/Unit/Controller/Store/RedirectTest.php]: Please `use Magento\Store\Model\StoreManagerInterface;` in all similar class and use the Alias.

File: app/code/Magento/Store/Test/Unit/Controller/Store/RedirectTest.php [hunk 4]
```diff
@@ -0,0 +1,140 @@
...
+            ]
+        );
+    }
+
+    /**
+     * @return void
+     */
+    public function testExecute()
+    {
+        $storeToSwitchToCode = 'sv2';
```
Comment (edenduong) [inline app/code/Magento/Store/Test/Unit/Controller/Store/RedirectTest.php]: Please add the const `const STUB_STORE_TO_SWITCH_TO_CODE = 'sv2';` and `const STUB_DEFAULT_STORE_VIEW = 'default';`

File: app/code/Magento/Store/Test/Unit/Controller/Store/RedirectTest.php [hunk 5]
```diff
@@ -0,0 +1,140 @@
...
+ */
+class RedirectTest extends \PHPUnit\Framework\TestCase
+{
+    /**
+     * @var \Magento\Store\Controller\Store\SwitchAction
+     */
+    private $model;
+
+    /**
+     * @var StoreRepositoryInterface|\PHPUnit_Framework_MockObject_MockObject
```
Comment (edenduong) [inline app/code/Magento/Store/Test/Unit/Controller/Store/RedirectTest.php]: `PHPUnit_Framework_MockObject_MockObject` should be `PHPUnit\Framework\MockObject\MockObject`

**Review by edenduong — CHANGES_REQUESTED** [review]
Hi. Please add `use ...` to your code and add some `const`. Thanks.

**Comment by TobiasCodeNull** [comment]
Cleaned up code and added Const values.

**Comment by edenduong** [comment]
> Cleaned up code and added Const values.

Hi @TobiasCodeNull. Please resolve the confict code. And please use the alias in the DocBlock also.

**Comment by engcom-Charlie** [comment]
Hi, @edenduong I'll take care of this.

**Comment by TobiasCodeNull** [comment]
I'll leave this to @engcom-Charlie then

**Review by edenduong — APPROVED** [review]

**Comment by magento-engcom-team** [comment]
Hi @edenduong, thank you for the review.
[ENGCOM-6869](https://jira.corp.magento.com/browse/ENGCOM-6869) has been created to process this Pull Request

**Comment by magento-engcom-team** [comment]
@TobiasCodeNull thank you for contributing. Please accept Community Contributors team invitation [here](https://github.com/orgs/magento/invitation) to gain extended permissions for this repository.

**Comment by engcom-Delta** [comment]
:heavy_check_mark: QA passed

**Comment by viktym** [comment]
For Magento versions 2.3.0 - 2.3.5 patch is available in [Magento Quality Patches](https://devdocs.magento.com/guides/v2.4/comp-mgr/patching.html#mqp) package (MQP)

**MQP Installation**
`$ composer require magento/quality-patches`

**Applying a Patch**
`$ ./vendor/bin/magento-patches apply MDVA-30107`

**Patch Details**
https://support.magento.com/hc/en-us/articles/360049869051

**Pull request merged**